File: examples/cpp/source/pca_transform/pca_transform_dense_batch.cpp
/* file: pca_transform_dense_batch.cpp */
/*******************************************************************************
* Copyright 2014-2019 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    C++ example of principal component analysis transformation(PCA)
!    in the batch processing mode with reduction
!
!******************************************************************************/

/**
 * <a name="DAAL-EXAMPLE-CPP-PCA_TRANSFORM_DENSE_BATCH"></a>
 * \example pca_transform_dense_batch.cpp
 */

#include "daal.h"
#include "service.h"

using namespace std;
using namespace daal;
using namespace daal::algorithms;

/* Input data set parameters */
const string dataFileName = "../data/batch/pca_transform.csv";
const size_t nVectors = 4;
const size_t nComponents = 2;

int main(int argc, char *argv[])
{
    checkArguments(argc, argv, 1, &dataFileName);
    FileDataSource<CSVFeatureManager> dataSource(dataFileName, DataSource::doAllocateNumericTable,
        DataSource::doDictionaryFromContext);
    dataSource.loadDataBlock(nVectors);

    /* Create an algorithm for principal component analysis using the SVD method */
    pca::Batch<double, pca::svdDense> pcaAlgorithm;

    pcaAlgorithm.input.set(pca::data, dataSource.getNumericTable());

    pcaAlgorithm.parameter.resultsToCompute = pca::mean | pca::variance | pca::eigenvalue;

    /* Compute results of the PCA algorithm*/
    pcaAlgorithm.compute();

    pca::ResultPtr pcaResult = pcaAlgorithm.getResult();

    /* Output basis, eigenvalues and mean values*/
    printNumericTable(pcaResult->get(pca::eigenvalues), "Eigenvalues:");
    printNumericTable(pcaResult->get(pca::eigenvectors), "Eigenvectors:");

    KeyValueDataCollectionPtr resultCollection = pcaResult->get(pca::dataForTransform);
    NumericTablePtr eigenvaluesT = NumericTable::cast((*resultCollection)[pca::eigenvalue]);
    if (eigenvaluesT.get() != NULL)
        printNumericTable(eigenvaluesT, "Eigenvalues kv:");

    NumericTablePtr meansT = NumericTable::cast((*resultCollection)[pca::mean]);
    if (meansT.get() != NULL)
        printNumericTable(meansT, "Means kv:");

    NumericTablePtr variancesT = NumericTable::cast((*resultCollection)[pca::variance]);
    if (variancesT.get() != NULL)
        printNumericTable(variancesT, "Variances kv:");


    /* Apply transform with whitening because means and eigenvalues are provided*/
    pca::transform::Batch<float> pcaTransform(nComponents);
    pcaTransform.input.set(pca::transform::data, dataSource.getNumericTable());
    pcaTransform.input.set(pca::transform::eigenvectors, pcaResult->get(pca::eigenvectors));

    pcaTransform.input.set(pca::transform::dataForTransform, pcaResult->get(pca::dataForTransform));

    pcaTransform.compute();

    /* Output transformed data */
    pca::transform::ResultPtr pcaTransformResult = pcaTransform.getResult();
    printNumericTable(pcaTransformResult->get(pca::transform::transformedData), "Transformed data:");

    return 0;
}

File: examples/python/source/pca/transform/pca_transform_dense_batch.py
# file: pca_transform_dense_batch.py
#===============================================================================
# Copyright 2014-2019 Intel Corporation.
#
# This software and the related documents are Intel copyrighted  materials,  and
# your use of  them is  governed by the  express license  under which  they were
# provided to you (License).  Unless the License provides otherwise, you may not
# use, modify, copy, publish, distribute,  disclose or transmit this software or
# the related documents without Intel's prior written permission.
#
# This software and the related documents  are provided as  is,  with no express
# or implied  warranties,  other  than those  that are  expressly stated  in the
# License.
#===============================================================================

#
# !  Content:
# !    Python example of PCA transformation algorithm.
# !*****************************************************************************

#
## <a name="DAAL-EXAMPLE-PY-PCA_TRANSFORM_DENSE_BATCH"></a>
## \example pca_transform_dense_batch.py
#

import os
import sys
import numpy as np

import daal.algorithms.pca as pca
import daal.algorithms.pca.transform as pca_transform
from daal.data_management import DataSourceIface, FileDataSource

utils_folder = os.path.realpath(os.path.abspath(os.path.dirname(os.path.dirname(os.path.dirname(__file__)))))
if utils_folder not in sys.path:
    sys.path.insert(0, utils_folder)
from utils import printNumericTable
from daal.data_management import NumericTable
# Input data set parameters
datasetName = os.path.join('..', 'data', 'batch', 'pca_transform.csv')

if __name__ == "__main__":

    # Retrieve the input data
    dataSource = FileDataSource(datasetName,
                                DataSourceIface.doAllocateNumericTable,
                                DataSourceIface.doDictionaryFromContext)
    dataSource.loadDataBlock()
    data = dataSource.getNumericTable()

    # Create an algorithm
    algorithm = pca.Batch(fptype=np.float64,method=pca.svdDense)

    # Set the algorithm input data
    algorithm.input.setDataset(pca.data, data)

    # Set the algorithm normalization parameters (mean and variance)
    # to be exported for transform and whitening parameter (eigenvalue)
    # If whitening is not required eigenvalues should be removed
    # The eigenvalues would be calculated in pca.eigenvalues table of result
    # but would not be passed to dataForTranform collection
    # algorithm.paramter.resultsToCompute = (pca.mean | pca.variance | pca.eigenvalue)

    algorithm.parameter.resultsToCompute = pca.mean | pca.variance | pca.eigenvalue;

    # Compute PCA
    res = algorithm.compute()
    # Output basis, eigenvalues and mean values
    printNumericTable(res.get(pca.eigenvalues), "Eigenvalues:")
    printNumericTable(res.get(pca.eigenvectors), "Eigenvectors:")

    eigenvaluesT = res.get(pca.eigenvalues)
    printNumericTable(eigenvaluesT, "Eigenvalues kv:")

    meansT = res.get(pca.means)
    printNumericTable(meansT, "Means kv:")

    #eigenvaluesT = res.getCollection(pca.eigenvalue)
    variancesT = res.get(pca.variances)
    printNumericTable(variancesT, "Variances kv:")

    # Create an algorithm
    tralgorithm = pca_transform.Batch(fptype=np.float64)

    # Set lower and upper bounds for the algorithm
    tralgorithm.parameter.nComponents = 2

    # Set an input object for the algorithm
    tralgorithm.input.setTable(pca_transform.data, data)

    # Set an input object for the eigenvectors
    tralgorithm.input.setTable(pca_transform.eigenvectors, res.get(pca.eigenvectors))

    # Set an input object for the eigenvectors
    tralgorithm.input.setCollection(pca_transform.dataForTransform, res.getCollection(pca.dataForTransform))

    # Compute PCA transformation function
    trres = tralgorithm.compute()

    printNumericTable(trres.get(pca.transform.transformedData), "Transformed data:");
    #printNumericTable(data, "First rows of the input data:", 4)
    #printNumericTable(trres.get(pca_transform.transformedData), "First rows of the min-max normalization result:", 4)
